buildSrc/src/main/java/gradlebuild/PublishPlugin.java

```diff
@@ -3,31 +3,122 @@
+            })));
+    }
+
+    private void configureRepositories(Project project) {
+        PublishRepositoryCredentials credentials = configureCredentials(project);
         project.getExtensions().configure(
             PublishingExtension.class,
             extension -> {
                 RepositoryHandler repositories = extension.getRepositories();
-                Stream.of(VersionDetails.ReleaseRepository.values())
-                    .filter(it -> it.getType() == VersionDetails.RepositoryType.Maven)
-                    .forEach(repository -> repositories.maven(repo -> {
-                        repo.setUrl(repository.getUrl());
-                        repo.setName(repository.name());
-                        repo.credentials(passwordCredentials -> {
-                            passwordCredentials.setUsername(credentials.getUserName());
-                            passwordCredentials.setPassword(credentials.getApiKey());
-                        });
-                    }));
-            });
+                configureLocalPublishRepository(() -> getLocalRepoDirectory(project), repositories);
+                configureRemotePublishRepositories(credentials, repositories);
+            }
+        );
+    }
+
+    private void configureLocalPublishRepository(Callable<File> localRepoDir, RepositoryHandler repositories) {
+        repositories.maven(repo -> {
+            repo.setName(LOCAL_FILE_REPOSITORY_NAME);
+            repo.setUrl(localRepoDir);
+        });
+    }
+
+    private void configureRemotePublishRepositories(PublishRepositoryCredentials credentials, RepositoryHandler repositories) {
+        Stream.of(VersionDetails.ReleaseRepository.values())
+            .forEach(repository -> repositories.maven(repo -> {
+                repo.setUrl(repository.getUrl());
+                repo.setName(repository.name());
+                repo.credentials(passwordCredentials -> {
+                    passwordCredentials.setUsername(credentials.getUserName());
+                    passwordCredentials.setPassword(credentials.getApiKey());
+                });
+            }));
+    }
+
+    private PublishRepositoryCredentials configureCredentials(Project project) {
+        PublishRepositoryCredentials credentials = project.getExtensions().create("publishRepository", PublishRepositoryCredentials.class);
+        if (project.hasProperty(PUBLISH_USER_NAME_PROPERTY)) {
+            credentials.setUserName(project.property(PUBLISH_USER_NAME_PROPERTY).toString());
+        }
+        if (project.hasProperty(PUBLISH_API_KEY_PROPERTY)) {
+            credentials.setApiKey(project.property(PUBLISH_API_KEY_PROPERTY).toString());
+        }
+        return credentials;
+    }
+
+    public static String getArchivesBaseName(Project project) {
+        BasePluginConvention convention = project.getConvention().findPlugin(BasePluginConvention.class);
+        return convention != null ? convention.getArchivesBaseName() : project.getName();
+    }
```

Comment:
When is `convention` `null` here? Will this be better with `BasePluginExtension` being introduced with Gradle 7.1?

Reply:
I think the `root` project does not apply the `BasePlugin`, that is when the convention will be null. I don't remember why I introduced it back then, I suppose it was necessary. I don't know if this is still necessary though.

Reply:
I removed the `null` handling. I don't know about `BasePluginExtension` or what would be different there.